File: Task2/c++/main.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int main() {
    cout << "Количество колец: ";
    int N;
    cin >> N;

    string rings;
    cout << "Введите строку с кольцами: " << endl;
    cin >> rings;

    // Проверка длины строки
    if (rings.length() != 2 * N) {
        cout << "Ошибка в воде строки" << endl;
        return 1;
    }

    // Массив для хранения цветов
    int colors[10] = {0};

    // Обработка каждого кольца
    for (int i = 0; i < N; i++) {
        char color = rings[2 * i];
        char rodChar = rings[2 * i + 1];

        // Проверка цвета
        if (color != 'r' && color != 'g' && color != 'b') {
            cout << "Ошибка в воде строки " << endl;
            return 1;
        }

        // Проверка номера стержня
        if (rodChar < '0' || rodChar > '9') {
            cout << "Ошибка в воде строки" << endl;
            return 1;
        }

        int index = rodChar - '0';

        // Сложение битов
        if (color == 'r') colors[index] += 1;
        else if (color == 'g') colors[index] += 2;
        else if (color == 'b') colors[index] += 4;
    }

    vector<int> result;
    for (int i = 0; i < 10; i++) {
        if (colors[i] == 7) {
            result.push_back(i);
        }
    }

    cout << "Количество стержней с кольцами всех трёх цветов: " 
         << result.size() << endl;

    if (result.empty()) {
        cout << "Нет стержней с тремя цветами" << endl;
    } else {
        cout << "Номера этих стержней: ";
        for (int i = 0; i < result.size(); i++) {
            cout << result[i] << " ";
        }
        cout << endl;
    }

    return 0;
}

File: Task2/rust/main/src/main.rs
use std::io;

fn main() {
    // Чтение количества колец
    let mut input = String::new();
    println!("Количество колец: ");
    io::stdin().read_line(&mut input).unwrap();
    let n: usize = input.trim().parse().unwrap();

    // Чтение строки с кольцами
    println!("Строка с кольцами: ");
    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    let rings = input.trim();

    // Массив для хранения цветов на 10 стержнях
    let mut colors = [0u8; 10];

    // Обработка каждого кольца
    for i in 0..n {
        let color = rings.chars().nth(i * 2).unwrap();
        let index = rings.chars().nth(i * 2 + 1).unwrap().to_digit(10).unwrap() as usize;

        if color == 'r' { colors[index] = colors[index] | 1 }
        else if color == 'g' { colors[index] = colors[index] | 2 }
        else if color == 'b' { colors[index] = colors[index] | 4 }
    }

    // Подсчёт стержней с кольцами всех цветов
    let mut result = Vec::new();
    let mut count = 0;
    for i in 0..10 {
        if colors[i] == 7 {
            result.push(i);
            count = count + 1
        }
    }

    println!("Количество стержней с кольцами всех трёх цветов: {}", count);

    if count == 0 {
        println!("Нет стержней с тремя цветами");
    } else {
        print!("Номера этих стержней: ");
        for x in &result { print!("{} ", x) }
        println!()
    }
}